plugins/adrv9009_adv.c

```diff
@@ -1013,107 +1030,164 @@ static int adrv9009adv_handle_driver(const char *attrib, const char *value)
+	                         G_CALLBACK(bist_tone_cb), plugin);
 
 	g_builder_connect_signal(builder, "tx_nco_enable", "toggled",
-	                         G_CALLBACK(bist_tone_cb), builder);
+	                         G_CALLBACK(bist_tone_cb), plugin);
 
 	g_builder_connect_signal(builder, "notebook", "switch-page",
 	                         G_CALLBACK(change_page_cb),
 	                         GTK_WIDGET(gtk_builder_get_object(builder, "initialize")));
 
-	can_update_widgets = true;
+	priv->can_update_widgets = true;
 
 	return adrv9009adv_panel;
 }
 
-static void update_active_page(gint active_page, gboolean is_detached)
+static void update_active_page(struct osc_plugin *plugin, gint active_page, gboolean is_detached)
 {
-	this_page = active_page;
-	plugin_detached = is_detached;
+	plugin->priv->this_page = active_page;
+	plugin->priv->plugin_detached = is_detached;
 }
 
-static void save_profile(const char *ini_fn)
+static void save_profile(struct osc_plugin *plugin, const char *ini_fn)
 {
 	FILE *f = fopen(ini_fn, "a");
 
 	if (f) {
-		save_to_ini(f, THIS_DRIVER, dev, adrv9009_adv_sr_attribs,
-		            ARRAY_SIZE(adrv9009_adv_sr_attribs));
+		save_to_ini(f, plugin->priv->plugin_name, plugin->priv->dev,
+			adrv9009_adv_sr_attribs, ARRAY_SIZE(adrv9009_adv_sr_attribs));
 		fclose(f);
 	}
 }
 
-static void context_destroy(const char *ini_fn)
+gpointer copy_gchar_array(gconstpointer src, gpointer data)
 {
-	save_profile(ini_fn);
-	osc_destroy_context(ctx);
+	return (gpointer)g_strdup(src);
 }
 
-static bool adrv9009adv_identify(void)
+static void context_destroy(struct osc_plugin *plugin, const char *ini_fn)
 {
-	/* Use the OSC's IIO context just to detect the devices */
-	struct iio_context *osc_ctx = get_context_from_osc();
+	save_profile(plugin, ini_fn);
 
-	dev = iio_context_find_device(osc_ctx, PHY_DEVICE);
+	size_t n = 0;
+	for (; n < plugin->priv->sr_attribs_count; n++) {
+		g_free(plugin->priv->sr_attribs[n]);
+	}
+	g_free(plugin->priv->sr_attribs);
 
-	return !!dev && iio_device_get_debug_attrs_count(dev);
+	osc_plugin_context_free_resources(&plugin->priv->plugin_ctx);
+	
+	osc_destroy_context(plugin->priv->ctx);
+	
+	g_free(plugin->priv);
 }
 
-struct osc_plugin plugin = {
-	.name = THIS_DRIVER,
-	.identify = adrv9009adv_identify,
-	.init = adrv9009adv_init,
-	.handle_item = adrv9009adv_handle,
-	.handle_external_request = handle_external_request,
-	.update_active_page = update_active_page,
-	.save_profile = save_profile,
-	.load_profile = load_profile,
-	.destroy = context_destroy,
-};
+struct osc_plugin * create_plugin(struct osc_plugin_context *plugin_ctx)
+{
+	struct osc_plugin *plugin= g_new0(struct osc_plugin, 1);
+
+	if (!plugin_ctx ) {
+		fprintf(stderr, "Cannot create plugin because an invalid plugin context was provided\n");
+		return NULL;
+	}
+
+	plugin->priv = g_new0(struct plugin_private, 1);
+	plugin->priv->plugin_ctx.plugin_name = g_strdup(plugin_ctx->plugin_name);
```

Comment:
there doesn't seem to be an equivalent free for this; or i missed it

Reply:
There is. The osc_plugin_context_free_resources() takes care of this.